### PR TITLE
fix #87: wait for fresh chain 15x faster

### LINKED ISSUE #87: downloading fresh chain from network takes too long even when everything is working correctly
This has been mentioned in #1.
Some rough metrics, calculated on my head and without using a timing device:
- Chain download speed "feels like" about 10 blocks per second or less
- That translates to over 1 minute for each day that the chain is old, to catch up
- The latest chain.zip is from mid-May
- So we're looking at 1-2 hours before a fresh install can start mining.

### PR DESCRIPTION
**Pull request by atramos**
Implemented a profiler and then optimized the slow parts.

### CODE REVIEW

**Review by lk16 — CHANGES_REQUESTED** [review]
The faster chain loading from peers is working, that is very nice.

However I notice that it wants to download the chain from peers again when I run the script for a second time.

The chain folder does contain all the blocks until 97100. 
At the time of writing the chain height is 97196, so if I understand the comments correctly, this is expected behaviour.

I also cloned your branch in another folder and tried again, just to make sure that my dev folder is not broken somehow. But I got the same results again.

**Comment by atramos** [comment]
Thanks for the feedback. I've switched this PR back to Draft until the comments are addressed.

File: skepticoin/networking/remote_peer.py [hunk 1]
```diff
@@ -463,43 +435,61 @@ def handle_data_message_received(self, header: MessageHeader, message: DataMessa
...
+
+                    valid_blocks = self.blocks_waiting_for_validation.copy()
+                    self.blocks_waiting_for_validation.clear()
+
+                    def write_pending_blocks() -> None:  # very slow, but ok to run in a separate thread!
+                        for valid_block in valid_blocks:
+                            self.local_peer.disk_interface.save_block(valid_block)
+                    Thread(target=write_pending_blocks, args=[]).start()
+
+                except Exception:
```
[the remark this answers is not in the document]
Comment (atramos) [reply skepticoin/networking/remote_peer.py:466]: > However I notice that it wants to download the chain from peers again when I run the script for a second time.

I was able to reproduce this. It's fixed in c54ff65900f48e98854f0c6a6645bb05464372cf

[the remark this answers is not in the document]
Comment (atramos) [reply skepticoin/networking/remote_peer.py:466]: > If we only want to catch block validation issues here and nothing else, this `try` block should not have anything else.
> I suggest moving the rest of the `try` block after the exception handling.

Fixed in d7a48dd1d10679c3cf6b1ed0e4cf20294fe669db

**Comment by atramos** [comment]
As part of the fixes in c54ff65900f48e98854f0c6a6645bb05464372cf, this PR also makes the `chain.cache` the primary block store and phases out the use of a single-file for every block in the chain.

File: skepticoin/scripts/utils.py [hunk 2]
```diff
@@ -63,57 +63,64 @@ def read_chain_from_disk() -> CoinState:
...
+                if os.path.getsize(f"chain/{filename}") == 0:
+                    print("Stopping at empty block file: %s" % filename)
+                    break
 
-    if fresher_chain:
-        write_chain_cache_to_disk(coinstate)
+                with open(Path("chain") / filename, 'rb') as f:
+                    try:
+                        block = Block.stream_deserialize(f)
+                    except Exception as e:
```
Comment (lk16) [inline skepticoin/scripts/utils.py:109]: Time to implement some proper exception handling, see #91. Not in this PR of course.

**Review by lk16 — APPROVED** [review]
I'll approve this, but if you merge this without changing the testcases, please create a new issue for it.

File: skepticoin/scripts/utils.py [hunk 3]
```diff
@@ -63,57 +63,64 @@ def read_chain_from_disk() -> CoinState:
...
+            coinstate = CoinState.zero()
 
-                    coinstate = coinstate.add_block_no_validation(block)
+    rewrite = False
 
-    else:
-        coinstate = CoinState.zero()
-        height = 0
+    if os.path.isdir('chain'):
+        # the code below is no longer needed by normal users, but some old testcases still rely on it:
```
[the remark this answers is not in the document]
Comment (atramos) [reply skepticoin/scripts/utils.py:92]: Created issue #93.

File: skepticoin/scripts/utils.py [hunk 4]
```diff
@@ -28,12 +32,8 @@ def __init__(self, *args: Any, **kwargs: Any):
...
 
 def create_chain_dir() -> None:
-    if not os.path.exists('chain'):
-        print("Pre-download blockchain from trusted source to 'blockchain-master'")
-        with urllib.request.urlopen("https://github.com/skepticoin/blockchain/archive/refs/heads/master.zip") as resp:
-            with open('chain.zip', 'wb') as outfile:
-                outfile.write(resp.read())
-        os.mkdir('chain')
+    if os.path.exists('chain'):
+        print('Your ./chain/ directory is no longer needed: please delete it to stop this reminder.')
```
[the remark this answers is not in the document]
Comment (atramos) [reply skepticoin/scripts/utils.py:36]: I'll rename create_chain_dir to check_chain_dir before merging this PR. I'll create an issue about deleting the obsolete files.

File: skepticoin/scripts/utils.py (hunk 2 above)
Comment (atramos) [reply skepticoin/scripts/utils.py:109]: I agree!

**Pull request merged**